libcasr/src/ubsan.rs

```diff
@@ -0,0 +1,256 @@
+//! UndefinedBehaviorSanitizer module implements `Severity` and `CrashLineExt` traits for UndefinedBehaviorSanitizer warnings.
+use crate::asan::AsanStacktrace;
+use crate::severity::Severity;
+use crate::stacktrace::{CrashLine, CrashLineExt};
+use crate::stacktrace::{ParseStacktrace, Stacktrace};
+
+use crate::error::*;
+use crate::execution_class::ExecutionClass;
+use regex::Regex;
+
+use std::fmt;
+
+/// Structure provides an interface for parsing ubsan runtime error message.
+#[derive(Clone)]
+pub struct UbsanWarning {
+    message: String,
+}
+
+impl fmt::Debug for UbsanWarning {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.message)
+    }
+}
+
+impl UbsanWarning {
+    pub fn extract_stacktrace(&self) -> Result<Vec<String>> {
+        AsanStacktrace::extract_stacktrace(&self.message)
+    }
+    pub fn parse_stacktrace(entries: &[String]) -> Result<Stacktrace> {
+        AsanStacktrace::parse_stacktrace(entries)
+    }
+    /// Get ubsan runtime error message as a line vector
+    pub fn ubsan_report(&self) -> Vec<String> {
+        self.message.split('\n').map(|s| s.to_string()).collect()
+    }
+}
+
+impl Severity for UbsanWarning {
+    fn severity(&self) -> Result<ExecutionClass> {
+        let message: Vec<String> = self
+            .message
+            .split('\n')
+            .map(|l| l.trim_end().to_string())
+            .collect();
+        // Get description (from first line)
+        let description = message.first().unwrap();
+        let re = Regex::new(r#".+: runtime error: (.+)"#).unwrap();
+        let Some(cap) = re.captures(description) else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse error description: {description}"
+            )));
+        };
+        let description = cap.get(1).unwrap().as_str().to_string();
+        // Get short description (from last line)
+        let short_description = message.last().unwrap();
+        let re = Regex::new(r#"SUMMARY: UndefinedBehaviorSanitizer: (\S+)"#).unwrap();
+        let Some(cap) = re.captures(short_description) else {
+            return Err(Error::Casr(format!(
+                "Couldn't parse ubsan summary: {short_description}"
+            )));
+        };
+        let short_description = cap.get(1).unwrap().as_str().to_string();
+
+        Ok(ExecutionClass::new((
+            "NOT_EXPLOITABLE",
+            &short_description,
+            &description,
+            "",
+        )))
+    }
+}
+
+impl CrashLineExt for UbsanWarning {
+    fn crash_line(&self) -> Result<CrashLine> {
+        let message: Vec<String> = self
+            .message
+            .split('\n')
+            .map(|l| l.trim_end().to_string())
+            .collect();
+        let mut re = Regex::new(r#".+ (.+):(\d+):\d+"#).unwrap();
+        // If there is no stacktrace use crashline from first string
+        // May be not absolute
+        // Else use first string from stacktrace
```

Comment:
We'd better reuse stack frame parsing from `asan.rs` to extract crash line, because lots of corner cases are handled there.

Reply:
You can move part of code from `asan:: parse_stacktrace` to `asan::parse_stacktrace_entry` and call the later one from your code.

Reply:
Done